### ISSUE TITLE
Error while trying to load Grace VST (Windows freeware)

### ISSUE THREAD

**Issue by mestaritonttu** [LMMS/lmms#2783]
[Grace sampler](http://onesmallclue.com/plugin/grace/) cannot be loaded. RemoteVstPlugin.exe "An error occurred in the application" dialog pops up.
Win 7 64-bit. LMMS 1.1.90
Reproduced on two different computers.

**Comment by tresf** [comment]
Please see the response in https://github.com/LMMS/lmms/issues/2784#issuecomment-219413242.  We don't have anyone actively debugging VSTs for LMMS currently (maintaining VST support has been a struggle in recent history).  Please do not be offended if this bug report as well as your other one are consolidated into a single report for better tracking.

**Comment by jasp00** [comment]
The 32-bit version does not work in Linux.

Grace is a VST 2.4 plug-in, while LMMS advertises VST 2.3. I am not sure whether this bug can be fixed. VST 2.4 introduced 64-bit capabilities and there is 64-bit support in LMMS, so I do not know how much VST 2.4 support there is. I am worried about the available specification at http://asseca.org/, how did they get the information? AFAIK, Steinberg never released the specification without a restricted agreement.

There is a VST 2.4 host, jVSTwRapper, [linked by Yvan Grabit](http://ygrabit.steinberg.de/~ygrabit/public_html/index.html), but I do not believe that confirms a permission from Steinberg to embed the specification in Java (if that is what it does). Besides, the [homepage](http://jvstwrapper.sourceforge.net/) has a link to the VST 2.0 specification that should not be there. We could ask _daniel309_ if Steinberg's license allows him to do that, but I believe [it does not](http://www.dith.it/listing/vst_stuff/vstsdk2.4/doc/html/license.html).

**Comment by mestaritonttu** [comment]
Other 64-bit VST plugins load and play fine (2.4 was the first version of the SDK with 64-bit support). Tested were Helm, OBXD, Triple Cheese, TAL-NoiseMaker and TyrellN6.

**Comment by musikBear** [comment]
I *believe* it has to do with the build-method. If the VST is created with a simplified speed development kit, it fails. Also (imho) there are a **lot** of VSTs that *does work* with lmms, in fact the majority of *instrumets* does. Weather it is sensible to hack in the code for a single one that does *not* work, i am really not sure.
The situation is different for VSTs *effect-plugins*. Here there are more fails than successes, and that is imho, more important to solve, because several suites of VST'effects are hugely superior to the ladspa collection.

**Comment by tresf** [comment]
> I believe it has to do with the build-method. If the VST is created with a simplified speed development kit, it fails.

Can you provide a link to this information?

**Comment by musikBear** [comment]
http://www.synthedit.com/
is such a build-kit, and i think i have seen others.
> Can you provide a link to this information?

'Information' is, i think a tall order. I wrote *I believe* that is just a 'hunch'
After that a hypothesis would have to expressed.
Actual tests needed to be preformed, and finally a conclusion could be made. for now its a educated guess, based on a hunch.

**Comment by Umcaruje** [comment]
@musikBear I know T-Force Alpha Plus was made in synthedit, and it is reported to work in LMMS just as well as other synths. 

>Also, the important part in my input was not the kits, but wheater or not it is sensible to make code changes because one or a few specific VSTs instruments fails in lmms, while there at the same time are 100 that does work. 

Well, there is not a few of those instruments, there are a lot of VST's that perform badly in LMMS, in one way or another.

**Comment by liushuyu** [comment]
>Besides, the homepage has a link to the VST 2.0 specification that should not be there. We could ask daniel309 if Steinberg's license allows him to do that, but I believe it does not.

@jasp00 I found what may worry you in that licence:
>You cannot re-work or otherwise pass this technology off as your own.

Oh, well, he **did** embed that in Java, **but** the provided SDK is integrated in Java with JNI (The VST part is still in native code provided by SDK, he didn't re-implement the whole VST technology), which in this case, I think he bypassed something in this part. 

The licence said:

>You can make VST Plug-Ins or VST Host Applications and distribute them worldwide.

His wrapper should be considered as VST Host App. And this is explicitly permitted in the license.

What I concern about is if we adopt his wrapper, we will introduce new dependencies (Java, JDK and more) and cares for Java code...

**Comment by grejppi** [comment]
> What I concern about is if we adopt his wrapper, we will introduce new dependencies (JRE, JDK and more) and cares for Java code...

There is [Airwave](https://github.com/phantom-code/airwave) which wraps plugins made for Windows and presents them as if they were native.

If we implement native VST support (in #2616 after 1.2 ~~if it is ever released~~), dropping VeSTige later and encouraging users to use Airwave (or something else) could reduce dependencies and maintenance burden, and perhaps improve stability.

`</wildideas>`

**Comment by liushuyu** [comment]
>If we implement native VST support (in #2616 after 1.2 if it is ever released), dropping VeSTige later and encouraging users to use Airwave (or something else) could reduce dependencies and maintenance burden, and perhaps improve stability.

No, I think the native VST support plugin will still be called "VeSTige". (~~Hey, please, if name changed some users would not be able to find how to load VST in LMMS~~)

```XML
< -- ! Just Joking>
            </wildideas>
            ^
Error: No matching opening tag found in the input
```

**Comment by grejppi** [comment]
@liushuyu The idea is to have a single instrument plugin that presents a list of available instruments in external formats, be they VST or LV2 or whatever else. Just like how it is for effects already. The name "VeSTige" wouldn't be appropriate for something that supports everything. The existing VeSTige plugin however could be retained for backward compatibility for a transition period.

We're getting off-topic here, and all of what I just said is at the moment pure speculation, but I'd love to continue the discussion over at #2616. :+1:

**Comment by jasp00** [comment]
> What I concern about is if we adopt his wrapper, we will introduce new dependencies (JRE, JDK and more) and cares for Java code...

The idea is not to adopt jVSTwRapper, but to study the code to learn the VST API. The same idea goes with JUCE. Airwave may be more useful because there is a binary host (and a proper license).

**Comment by PhysSong** [comment]
Seems to be fixed in #3776, as well as #2784. @mestaritonttu Do you still have this issue with 1.2.0-rc4?

**Comment by mestaritonttu** [comment]
It works now, thanks!
